pkg/agent/route/route_windows.go

```diff
@@ -213,109 +220,116 @@ func (c *Client) DeleteRoutes(podCIDR *net.IPNet) error {
+		return nil
 	}
-	if found {
-		existingRoute := obj.(*util.Route)
-		if existingRoute.GatewayAddress.Equal(route.GatewayAddress) && existingRoute.RouteMetric == route.RouteMetric {
-			klog.V(2).InfoS("Service route already exists", "DestinationIP", route.DestinationSubnet,
-				"Gateway", route.GatewayAddress, "RouteMetric", route.RouteMetric)
-			return nil
+
+	var newClusterIPCIDR *net.IPNet
+	var err error
+	if curClusterIPCIDR != nil {
+		// If the route exists and its destination CIDR doesn't contain the ClusterIP, generate a new destination CIDR by
+		// enlarging the current destination CIDR with the ClusterIP.
+		if newClusterIPCIDR, err = util.ExtendCIDRWithIP(curClusterIPCIDR, svcIP); err != nil {
+			return fmt.Errorf("enlarge the destination CIDR with an error: %w", err)
 		}
-		// Remove the existing route if gateway or metric is not as expected.
-		if err := util.RemoveNetRoute(existingRoute); err != nil {
-			return fmt.Errorf("failed to delete existing Service route entry, DestinationIP: %s, Gateway: %s, RouteMetric: %d, err: %v",
-				existingRoute.DestinationSubnet, existingRoute.GatewayAddress, existingRoute.RouteMetric, err)
+	} else {
+		// If the route doesn't exist, generate a new destination CIDR with the ClusterIP. Note that, this is the first
+		// ClusterIP since the route doesn't exist.
+		newClusterIPCIDR = util.NewIPNet(svcIP)
+	}
+
+	// Generate a route with the new destination CIDR and install it.
+	route := generateRoute(newClusterIPCIDR, gw, linkIndex, metric)
+	if err = util.ReplaceNetRoute(route); err != nil {
+		return fmt.Errorf("failed to install new ClusterIP route: %w", err)
+	}
+	// Store the new destination CIDR.
+	c.clusterIPv4CIDR = route.DestinationSubnet
+	klog.V(4).InfoS("Created a route to route the ClusterIP to Antrea gateway", "route", route, "ClusterIP", svcIP)
+
+	// Collect stale routes.
+	var staleRoutes []*util.Route
+	if curClusterIPCIDR != nil {
+		// If current destination CIDR is not nil, the route with current destination CIDR should be uninstalled since
+		// a new route with a newly calculated destination CIDR has been installed.
+		route.DestinationSubnet = curClusterIPCIDR
+		staleRoutes = []*util.Route{route}
+	} else {
+		// If current destination CIDR is nil, which means that Antrea Agent has just started, then all existing routes
+		// whose destination CIDR contains the first ClusterIP should be uninstalled, except the newly installed route.
+		// Note that, there may be multiple stale routes prior to this commit. When upgrading, all stale routes will be
+		// collected. After this commit, there will be only one stale route after Antrea Agent started.
+		routes, err := c.listIPRoutesOnGW()
+		if err != nil {
+			return fmt.Errorf("error listing ip routes: %w", err)
+		}
+		for _, rt := range routes {
+			if rt.GatewayAddress.Equal(gw) && !rt.DestinationSubnet.IP.Equal(svcIP) && rt.DestinationSubnet.Contains(svcIP) {
```

Comment:
The newly installed route in line 269 also satisfies this condition, so is it possible it is removed by mistake?

Reply:
If AntreaProxy only has one worker (goroutine), it cannot be removed by mistake since it is only used when the first Service is installed, and `curClusterIPCIDR` is nil.